docs/style-spec/expressions.md

```diff
@@ -0,0 +1,174 @@
+      'rgb',
+      // red is higher when feature.properties.temperature is higher
+      ["get", ["properties"], "temperature", "Number"],
+      0,
+      // blue is higher when feature.properties.temperature is lower
+      ["-", 100, ["get", ["properties"], "temperature", "Number"]]
+    ]
+  }
+}
+```
+
+
+## Zoom-dependent functions
+
+A <strong>zoom function</strong> is any function defined using an expression that includes a reference to `["zoom"]`.  Such functions allow the appearance of a map feature change with map’s zoom level. Zoom functions can be used to create the illusion of depth and control data density.  
+
+A zoom function must be of the following form (see [Curves](#Curves) below):
+
+```js
+{
+  "expression": [ "curve", interpolation, ["zoom"], ... ]
+}
+```
+
+(`["zoom"]` may not appear anywhere else in the expression.)
+
+
+### Example: a zoom-only function.
+
+```js
+{
+  "circle-radius": {
+    "expression": [
+      "curve", "linear", ["zoom"],
+      // zoom is 5 (or less) -> circle radius will be 1px
+      // zoom is 10 (or greater) -> circle radius will be 2px
+      5, 1, 10, 2
+    ]
+  }
+}
+```
+
+### Example: a zoom-and-property function
+
+Using property functions as the output value for one or more zoom stops allows
+the appearance of a map feature to change with both the zoom level _and_ the
+feature's properties.
+
+```js
+{
+  "circle-radius": {
+    "expression": [
+      "curve", "linear", ["zoom"],
+
+      // zoom is 0 and "rating" is 0 -> circle radius will be 0px
+      // zoom is 0 and "rating" is 5 -> circle radius will be 5px
+      0, [ "get", ["properties"], "rating", "Number" ],
+      // zoom is 20 and "rating" is 0 -> circle radius will be 4 * 0 = 0px
+      // zoom is 20 and "rating" is 5 -> circle radius will be 4 * 5 = 20px
+      10, [ "*", 4, [ "get", ["properties"], "rating", "Number" ] ]
+    ]
+  }
+}
+```
+
+## Property Expressions
+
+Property expressions are represented using a Lisp-like structured syntax tree.
+
+###Types
+
+Every expression evaluates to a value of one of the following types.
+
+- `Null`
+  - Literal: `null`
+- `String`
+  - Literal: `"my string value"`
+- `Number`
+  - Literal: `1729`
+- `Boolean`
+  - Literal: `true` or `false`
+- `Color`
+- `JSONObject`
+- `Array<T, N>`: an array of N values of type T
+  - Literal: `["array", v0, v1, ...]`
+- `Vector<T>`: a dynamically sized array of values of type T
+  - Literal: `["vector", v0, v1, v2, ... vN]`
+  - TODO: without type inference, 0-length arrays and vectors can't be typed
+- `Value`: A "variant" type representing the set of possible values retrievable from a feature's `properties` object (`Null | String | Number | Boolean | JSONObject | Vector<Value>`)
+- `Error`: a subtype of all other types. Used wherever an expression is unable to return the appropriate supertype. Carries diagnostic information such as an explanatory message and the expression location where the Error value was generated.
+
```

Comment:
@jfirebaugh Can you say more about why you think we’ll want these over strings?  specifically: are there cases where such types would be useful for checking what a _subexpression_‘s evaluates to? 

So far, the use cases I can think of for enums are for checking the _final_ output of an expression that's being used for an enum-typed style property like `symbol-placement`.  If it's true that this is the main reason for them, then the question is which is simpler/better: adding enums to the type system here, or treating the validation of style property values as a separate step?

Since there are, e.g., numeric properties like `*-opacity` that can't really be validated via types (at least not normal ones), seems like we're gonna have the extra validation step anyway, which makes me lean towards the latter.